Simplify Vite config

Rich-Harris opened the pull request:
Another stepping stone on the way to #7967 (which at this point is less important than the codebase improvements it's driving).

Right now the `vite` directory is a bit of a mess. The code is full of indirection, and the logic is split up in ways that are quite hard to follow. This simplifies everything.

The `vite-plugin-sveltekit-compile` plugin should be renamed `vite-plugin-sveltekit-serve`, though that would break the Storybook integration, which removes that plugin from the array.

benmccann commented on packages/kit/src/exports/vite/index.js:
```diff
@@ -333,17 +261,152 @@ function kit({ svelte_config }) {
+				optimizeDeps: {
+					exclude: ['@sveltejs/kit']
+				},
+				publicDir: ssr ? false : kit.files.assets,
```
this seems like something that belongs in the serve plugin and in fact it's being overridden there. does this one have any effect?

Rich-Harris replied:
errr.... not sure. but presumably the `ssr ? false : kit.files.assets` thing is for a reason. if it's not having an effect then perhaps something is broken? need to investigate

benmccann replied:
Ah, this plugin only runs in build mode and the serve plugin only runs if not in build mode, so it won't override this. I'll resolve this comment

benmccann commented on packages/kit/src/exports/vite/index.js:
```diff
@@ -240,16 +154,15 @@ function kit({ svelte_config }) {
 			// dev and preview config can be shared
 			/** @type {import('vite').UserConfig} */
-			const result = {
+			const new_config = {
```
`new_config` sounds a bit like it's the complete config. maybe `additional_config` to make it clearer that Vite will combine it with the provided config?

benmccann commented on packages/kit/src/exports/vite/index.js:
```diff
@@ -333,17 +261,152 @@ function kit({ svelte_config }) {
+					// don't use the default name to avoid collisions with 'static/manifest.json'
+					manifest: 'vite-manifest.json'
+				},
+				define: {
```
this kind of stuff seems like it belongs in the setup plugin

Rich-Harris replied:
why? the `define` values for build are different than for dev

benmccann replied:
I intended the setup plugin to contain all the setup regardless of whether it's dev or build - it can return different config depending on which command is run. Tools like Storybook will still need things like `__SVELTEKIT_DEV__` set appropriately in build mode.

Rich-Harris replied:
Do they? I think that stuff's only use by SvelteKit internals that shouldn't be imported in a Storybook context. Ww _can_ put it in `setup` but it'll increase complexity

benmccann replied:
Users may create components that reference [SvelteKit modules](https://kit.svelte.dev/docs/modules). If those modules use defines that are not available or set properly it will end up causing issues.

benmccann reviewed (commented):
Can we combine the build and serve plugins back to just a compile plugin? I don't know that we gain a lot by splitting it. This change will likely break Storybook and Histoire because they probably would need to ignore the build plugin as well, so at least it'd be good to send them PRs with the new plugin names first. But I'm not sure it's worth doing that. The serving part is already well separated by living in `configureServer` and `configurePreviewServer`. It also seems like this PR has some config in the build plugin that I'd put in the serve plugin and vice versa. I think we could just ignore that complication by keeping it a single plugin.

Rich-Harris commented on packages/kit/src/exports/vite/index.js:
```diff
@@ -618,13 +620,64 @@ function kit({ svelte_config }) {
+			const new_config = {
+				appType: 'custom',
+				base: svelte_config.kit.paths.base,
+				build: {
```
No, this is serve configuration

Rich-Harris commented:
Reverted it to a single plugin and moved the `define` stuff around, even though the code is more confusing now.

I really dislike the concessions we're making here. Something is deeply wrong with how these tools interact with each other. I don't know what the correct solution is, but mutating an array of plugins to remove one that has a specific name is definitely not it.

benmccann commented on packages/kit/test/apps/basics/test/client.test.js:
```diff
@@ -651,6 +651,21 @@ test.describe('Load', () => {
 	}
+
+	if (!process.env.DEV) {
+		test.skip('does not fetch __data.json if no server load function exists', async ({
```
do you mean to always skip this test?

Rich-Harris replied:
yes, we can't unskip it until we implement #7967. doing that _and_ reconfiguring the Vite plugin would have been a lot for one PR

benmccann reviewed (approved):
We should probably at some point provide an API that lets you set things up without adding the compile plugin, so that other tools don't have to remove it. I think splitting the plugins in the way that we are probably makes sense and is necessary in order to support other Vite-based tools that want to utilize SvelteKit components, but I agree having to remove a plugin is a bit gross
